### PR TITLE
[TTAHUB-948] update user roles

### PR DESCRIPTION
**Pull request by thewatermethod**
## Description of change

Convert the user roles from static data located in code to the Roles table. Update the roles table to have the correct verbiage. Update all the places in the code where the static role list was used.

## How to test

The places where the roles are used are:
- viewing and updating a user's roles in the admin screen
- the activity report landing page, to populate the author and collaborator columns, as well as in the roles filter
- the activity report form, to save the author and collaborator roles
- the approved activity report page
- the goals and objectives forms in both places, to populate the ObjectiveRoles

## Issue(s)

* https://ocio-jira.acf.hhs.gov/browse/TTAHUB-948

### Every PR

- [x] Meets issue criteria
- [x] JIRA ticket status updated
- [x] Code is meaningfully tested
- [ ] Meets accessibility standards (WCAG 2.1 Levels A, AA)
- [ ] API Documentation updated
- [ ] Boundary diagram updated
- [ ] Logical Data Model updated
- [ ] [Architectural Decision Records](https://adr.github.io/) written for major infrastructure decisions

### Production Deploy

- [ ] Staging smoke test completed

### After merge/deploy

- [ ] Update JIRA ticket status

### CODE REVIEW

File: src/migrations/20220808170215-unify-roles-table.js [hunk 1]
```diff
@@ -0,0 +1,120 @@
...
+                      "updatedAt"
+                  ) 
+              VALUES 
+                  (
+                      u.user_id, 
+                     (SELECT "id" from "Roles" WHERE u.user_role = "Roles"."fullName"),
+                     now(),
+                     now()
+                  );
+          END LOOP;
```
Comment (GarrettEHill) [inline src/migrations/20220808170215-unify-roles-table.js:96]:
```suggestion
     WITH
         "RolesForUser" as (
             SELECT
                 id "userId",
                 UNNEST("role")::text "roleName"
             FROM "Users"
         )
     INSERT INTO "UserRoles"
     (
          "userId",
          "roleId",
          "createdAt",
          "updatedAt"
      )
      SELECT DISTINCT
          rfu."userId",
          r.id "roleId",
          now() "createdAt",
          now() "updatedAt"
      FROM "RolesForUser" rfu
      JOIN "Roles" r
      ON rfu."roleName" = r."fullName"
      ORDER BY 1,2;
```

File: src/migrations/20220808170215-unify-roles-table.js [hunk 2]
```diff
@@ -0,0 +1,120 @@
...
+        },
+        createdAt: {
+          allowNull: false,
+          type: Sequelize.DATE,
+        },
+        updatedAt: {
+          allowNull: false,
+          type: Sequelize.DATE,
+        },
+      }, { transaction });
```
Comment (GarrettEHill) [inline src/migrations/20220808170215-unify-roles-table.js:72]:
```suggestion
      }, { 
          uniqueKeys: {
              userId_roleId_unique: {
                 fields: ['userId', 'roleId']
              }
          },
          transaction,
      });
```

Comment (thewatermethod) [reply src/migrations/20220808170215-unify-roles-table.js:72]: Will gladly take these suggestions, but would love to also hear a "why"

File: src/migrations/20220809151140-update-collaborator-roles.js [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+        DECLARE
+            u record;
+        BEGIN
+        FOR u IN SELECT "id" as collaborator_role_id, "role" as "collaborator_role" FROM "CollaboratorRoles"
+          LOOP               
+            UPDATE "CollaboratorRoles" SET "roleId" = (SELECT "id" from "Roles" WHERE u.collaborator_role = "Roles"."fullName") WHERE id = u.collaborator_role_id;
+          END LOOP; 
+          END;
+        $$
+        LANGUAGE plpgsql;
```
Comment (GarrettEHill) [inline src/migrations/20220809151140-update-collaborator-roles.js:52]: UPDATE "CollaboratorRoles" cr
SET "roleId" = r.id
FROM "Roles" r
WHERE cr.role = r."fullName";

File: src/migrations/20220809151140-update-collaborator-roles.js [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+        `,
```
Comment (GarrettEHill) [inline src/migrations/20220809151140-update-collaborator-roles.js:53]:
```suggestion
        `UPDATE "CollaboratorRoles" cr
        SET "roleId" = r.id
        FROM "Roles" r
        WHERE cr.role = r."fullName";`,
```

**Review by AdamAdHocTeam — COMMENTED** [review]
Found some things that need to be addressed.

1. The creator role selector doesn't show on 'Review and Submit' if they creator has more than one role. This causes the AR to fail to submit for review.

image

image

2. Creator and Collaborator roles don't show on the Alerts table

image

3. When viewing an approved report I don't see the collaborator roles.

image

---------------------------------------

Regarding the roles I will assume that we are keeping '**Contract Officer's Representitive**' as '**COR**'. Also '**Central Office: TTA and Comprehensive Services Division**' and '**Central Office: Other Divisions**' as just '**COR**'.

**Review by AdamAdHocTeam — APPROVED** [review]
Looks good.

My concerns have been addressed.

<html>
<body>


:cake:
--



</body>
</html>

**Pull request merged**